### PR TITLE
Interactive bluefog

### PR DESCRIPTION
**Pull request by kunyuan827**
add an example on resource allocation

### CODE REVIEW

File: examples/resource_allocation.ipynb [hunk 1]
```diff
@@ -0,0 +1,1495 @@
...
+    "\n",
+    "\n",
+    "3. How to setup topology, determine weights, and run neighbor_allreduce with BlueFog;\n",
+    "\n",
+    "\n",
+    "4. How to use BlueFog functions such as **bf.allreduce(), bf.neighbor_allreduce(), bf.load_topology(), bf.set_topology(), topology_util.GetRecvWeights()**, and **topology_util.ExponentialTwoGraph()**\n",
+    "\n",
+    "\n",
+    "5. How to implement EXTRA, Exact Diffusion, and Gradient Tracking with BlueFog"
+   ]
```
Comment (ybc1991) [inline examples/resource_allocation.ipynb:34]: Do not have extra newline in middle. Let the markdown give the index automaitically. 
You can check the preview, it is all 1. xx 1.xxx 1.xxx

File: examples/resource_allocation.ipynb [hunk 2]
```diff
@@ -0,0 +1,1495 @@
...
+   "outputs": [],
+   "source": [
+    "%%px\n",
+    "\n",
+    "# EXTRA\n",
+    "def EXTRA_one_recursion(y, gp, yp, ATA, A, b, x_sol, lr, self_weight, neighbor_weights):\n",
+    "    \n",
+    "    x = ATA.mm(A.T.mm(b) - y)\n",
+    "    g = -x\n",
+    "    yn = bf.neighbor_allreduce(2*y - yp, self_weight, neighbor_weights, name='y')\n",
```
[the remark this answers is not in the document]
Comment (kunyuan827) [reply examples/resource_allocation.ipynb:630]: Note that for EXTRA, we didn't set topology as set_topology(G, is_weighted=True). Instead, we use the exponential topology but use self and neighbor weights as W_bar. We shouldn't remove the weights in EXTRA code?

For exact diffusion, we use set_topology(G, is_weighted=True) and in that example, we removed self weight and neighbor_weights.

File: examples/resource_allocation.ipynb [hunk 3]
```diff
@@ -0,0 +1,1495 @@
...
+   "metadata": {},
+   "source": [
+    "\\begin{align*}\n",
+    "\\bar{w}_{ij}=\n",
+    "\\begin{cases}\n",
+    "(1+w_{ii})/2 & \\mbox{if $i=j$} \\\\\n",
+    "w_{ij}/2 & \\mbox{otherwise}\n",
+    "\\end{cases}\n",
+    "\\end{align*}"
+   ]
```
[the remark this answers is not in the document]
Comment (kunyuan827) [reply examples/resource_allocation.ipynb]: Try different ways to reformat the cell but don't work...

Don't know how to handle it...

[the remark this answers is not in the document]
Comment (ybc1991) [reply examples/resource_allocation.ipynb]: Ok, let it be. As long as jupyter notebook looks fine, it is ok.

File: examples/resource_allocation.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ybc1991) [reply examples/resource_allocation.ipynb:646]: My bad. I saw the new version and you also added the comment in Exact Diffusion to distinguish them. It looks good to me.

Please clear all the output and re-run the whole cells again. Make sure after the modification it still works fine and the execution number next to the cell is sequential.

File: examples/interactive_bluefog_helloworld.ipynb [hunk 4]
```diff
@@ -573,6 +573,98 @@
     "plt.show()"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Interrupt the hanged worker process\n",
+    "\n",
+    "It is quite common in the code dealing with the multiple machine communication that some unexpected bugs or deadlock caused the worker process(es) hanged forever. Recall the worker processes are the seperated environment from the notebook. Hence, the interrupt kernel in the tool bar of notebook can interrupt the code of execution in notebook but the worker processes are still running."
```
Comment (hanbinhu) [inline examples/interactive_bluefog_helloworld.ipynb:582]: - caused -> cause
- hanged -> hang
- are the separated -> are inside a separated
- in the tool bar of notebook -> in the toolbar of the notebook

File: bluefog/run/interactive_run.py [hunk 5]
```diff
@@ -151,6 +176,45 @@ def _maybe_kill_ipcontroller_process(profile):
...
+    if engine_pids is None:
+        return
+    for _, pid in engine_pids.items():
+        try:
+            os.kill(pid, signal.SIGINT)
+        except:
+            pass
+
+
+def interrupt_hanged_ipengine_processes(profile):
```
Comment (hanbinhu) [inline bluefog/run/interactive_run.py:192]: Didn't quite understand the logic of this function. Maybe add some comments to help, especially, this is a function exposed to users.

File: examples/interactive_bluefog_helloworld.ipynb [hunk 6]
```diff
@@ -573,6 +573,98 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's use sleep function to simulate that situation. If you run the following code and click the interrupt button afterwards, you will be able to execute other cells but the workers still continues to sleep, which will block the running of other requests."
```
Comment (hanbinhu) [inline examples/interactive_bluefog_helloworld.ipynb:589]: - Use some special function for `sleep`.
-  still continues -> still continue

File: examples/interactive_bluefog_helloworld.ipynb [hunk 7]
```diff
@@ -573,6 +573,98 @@
...
+    "print(\"Done\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Therefore, BlueFog provides a utility function `interrupt_hanged_ipengine_processes`. After running this functions, the previous sleep function will be interrupted so that you should be able to execute other functions afterwards.\n",
+    "\n",
+    "Note: `interrupt_hanged_ipengine_processes` is supported under the localhost mode only."
```
Comment (hanbinhu) [inline examples/interactive_bluefog_helloworld.ipynb:633]: - Try to highlight this line, and add additional comments in the code for this function.

File: examples/interactive_bluefog_helloworld.ipynb [hunk 8]
```diff
@@ -573,6 +573,98 @@
...
+    "# It is because the workers are still sleeping. The interruption in notebook\n",
+    "# does not interrupt the workers.\n",
+    "print(\"Done\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Therefore, BlueFog provides a utility function `interrupt_hanged_ipengine_processes`. After running this functions, the previous sleep function will be interrupted so that you should be able to execute other functions afterwards.\n",
```
Comment (hanbinhu) [inline examples/interactive_bluefog_helloworld.ipynb:631]: - this functions -> this function
- Use special font for `sleep`

File: examples/interactive_bluefog_helloworld.ipynb [hunk 9]
```diff
@@ -573,6 +573,98 @@
     "plt.show()"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Interrupt the hanged worker process\n",
```
Comment (hanbinhu) [inline examples/interactive_bluefog_helloworld.ipynb:580]: Consider change Interrupt to Restart or awake. Also change the API naming

**Review by hanbinhu — CHANGES_REQUESTED** [review]
I didn't look at the `resource_allocation.ipynb`, but checked the interruption part for the `interactive_run.py` and `interactive_bluefog_helloworld.ipynb`.

File: bluefog/run/interactive_run.py (hunk 5 above)
Comment (Bluefog-Lib) [reply bluefog/run/interactive_run.py:192]: Discussed offline. There was a misunderstanding here. "Interrupt" is fine.

File: examples/interactive_bluefog_helloworld.ipynb (hunk 4 above)
Comment (Bluefog-Lib) [reply examples/interactive_bluefog_helloworld.ipynb:582]: Done

File: examples/interactive_bluefog_helloworld.ipynb (hunk 6 above)
Comment (Bluefog-Lib) [reply examples/interactive_bluefog_helloworld.ipynb:589]: Done

File: examples/interactive_bluefog_helloworld.ipynb (hunk 7 above)
Comment (Bluefog-Lib) [reply examples/interactive_bluefog_helloworld.ipynb:633]: Done. Use * *

File: examples/interactive_bluefog_helloworld.ipynb (hunk 8 above)
Comment (Bluefog-Lib) [reply examples/interactive_bluefog_helloworld.ipynb:631]: Done

File: examples/interactive_bluefog_helloworld.ipynb (hunk 9 above)
Comment (Bluefog-Lib) [reply examples/interactive_bluefog_helloworld.ipynb:580]: Refer to the previous comment in the interrupt_hanged_processes .

File: examples/interactive_bluefog_helloworld.ipynb [hunk 10]
```diff
@@ -579,14 +579,14 @@
    "source": [
     "# Interrupt the hanged worker process\n",
     "\n",
-    "It is quite common in the code dealing with the multiple machine communication that some unexpected bugs or deadlock caused the worker process(es) hanged forever. Recall the worker processes are the seperated environment from the notebook. Hence, the interrupt kernel in the tool bar of notebook can interrupt the code of execution in notebook but the worker processes are still running."
+    "It is quite common in the code dealing with the multiple machine communication that some unexpected bugs or deadlock cause the worker process(es) hange forever. Recall the worker processes are inside a seperated environment from the notebook. Hence, the 'interrupt the kernel' button, i.e. black square box, in the toolbar of the notebook interrupts the code of execution in notebook but the worker processes are still running."
```
Comment (hanbinhu) [inline examples/interactive_bluefog_helloworld.ipynb:582]: - hange -> hang

**Review by hanbinhu — APPROVED** [review]

**Pull request merged**